src/mpi/comm/comm_split_type.c

```diff
@@ -73,6 +70,142 @@ int MPIR_Comm_split_type_node(MPIR_Comm * comm_ptr, int key, MPIR_Comm ** newcom
     goto fn_exit;
 }
 
+#ifdef HAVE_HWLOC
+
+struct shmem_info_table {
+    const char *val;
+    hwloc_obj_type_t hwloc_obj_type;
+    int depth;
+};
+
+static struct shmem_info_table shmem_info[] = {
+    {"socket", HWLOC_OBJ_PACKAGE, -1},
+    {"package", HWLOC_OBJ_PACKAGE, -1},
+    {"numanode", HWLOC_OBJ_NUMANODE, -1},
+    {"core", HWLOC_OBJ_CORE, -1},
+    {"hwthread", HWLOC_OBJ_PU, -1},
+    {"l1cache", HWLOC_OBJ_CACHE, 1},
+    {"l2cache", HWLOC_OBJ_CACHE, 2},
+    {"l3cache", HWLOC_OBJ_CACHE, 3},
+    {"\0", HWLOC_OBJ_TYPE_MAX}
+};
+
+static int comm_split_type_hwloc(const char *hardware_filter, int node_id, MPIR_Comm * comm_ptr,
+                                 MPIR_Comm ** newcomm_ptr, int key)
+{
+    int color = node_id;
+    int i;
+    int mpi_errno = MPI_SUCCESS;
+
+    if (!MPIR_Process.bindset_is_valid)
+        goto do_split;
+
+    for (i = 0; shmem_info[i].val; i++) {
+        if (!strcmp(shmem_info[i].val, hardware_filter)) {
+            hwloc_obj_type_t query_obj_type = shmem_info[i].hwloc_obj_type;
+            hwloc_obj_t obj_containing_cpuset;
+
+            /* we deal with cache objects separately since they
+             * require an additional depth query */
+            if (query_obj_type == HWLOC_OBJ_CACHE) {
+                obj_containing_cpuset =
+                    hwloc_get_cache_covering_cpuset(MPIR_Process.topology, MPIR_Process.bindset);
+                if (obj_containing_cpuset == NULL)
+                    goto do_split;
```

Comment:
In cases where the user asked us to split a certain way and we could not find the right object to do that, we should at least throw a warning saying so.